### ISSUE TITLE
Add ability to specify an overlay-style bezel

### ISSUE THREAD

**Issue by aaciii** [dosbox-staging/dosbox-staging#3558]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Windows 11

### Is your feature request related to a problem? Please describe.

No - not a problem per se

### Describe the solution you'd like

Would like the ability to specify (command line and/or config file) a transparent PNG file to overlay on the screen to create a border and/or decorate the sides of the screen (where a 16x9 display leaves black gaps as a 4x3 DOSbox session is running.

### Describe alternatives you've considered

I have implemented this with ReShade but that's a PITA to be honest.   Would be great to have it integrated like many other retro emulators.
![dosbox 2024-04-01 18-10-52](https://github.com/dosbox-staging/dosbox-staging/assets/80229871/a928d76e-494e-4c8b-8e40-c8696482d060)
![dosbox 2024-04-01 18-16-23](https://github.com/dosbox-staging/dosbox-staging/assets/80229871/cfd0a4f7-a09b-434c-b2e1-51ab9573e794)
![dosbox 2024-04-01 18-32-41](https://github.com/dosbox-staging/dosbox-staging/assets/80229871/4076595d-90ae-4c90-a235-143f0e0b54f6)
![dosbox 2024-04-01 18-34-57](https://github.com/dosbox-staging/dosbox-staging/assets/80229871/ac88d3cd-5dc3-4c9d-9f22-95ed7076efc9)

### Add any other context or screenshots about the feature request here.

Here's an example of a transparent PNG bezel.

![Layer](https://github.com/dosbox-staging/dosbox-staging/assets/80229871/1bffa552-81ce-4eaa-b998-b3faced53522)

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
I'm personally not a fan of such things, but if it doesn't cause major complications, we could add support for this later on. However, it's really low on my personal priority list.

**Comment by aaciii** [comment]
Sure thing.  I showed examples of a "realistic" bezel (NEC monitor) but obviously could be any image.  This is useful in providing on-screen instructions or embellishments for example (can eliminate the need to have a manual for a game).

![Screenshot 2024-04-01 223224](https://github.com/dosbox-staging/dosbox-staging/assets/80229871/d3dfa7d6-9ead-45b9-939c-6b126909be97)
![Screenshot 2024-04-01 223120](https://github.com/dosbox-staging/dosbox-staging/assets/80229871/acdec089-34b0-4d36-91c8-f4c2ba8d2f83)

**Comment by johnnovak** [comment]
> This is useful in providing on-screen instructions or embellishments for example (can eliminate the need to have a manual for a game).

...unless your game came with a 200+ page manual ;) But I get your point, this can be useful for various purposes.

**Comment by johnnovak** [comment]
I've made the executive decision to decline this feature, along with adopting the RetroArch shader pipeline.

We can discuss this again after the OSD, maybe, but let's consider the issue closed until then.

Related:

https://github.com/dosbox-staging/dosbox-staging/issues/100

**Comment by aaciii** [comment]
C'est la vie...   will look to another DOSbox project/fork I guess!

**Comment by Grounded0** [comment]
Next they ask for adding hard disk and floppy drive loading noises. Sure we could do all this but that would make the base product worse since it would take away resources from for example fixing game compatibility which DOSBox 0.74 had huge holes in.

**Comment by aaciii** [comment]
With respect, this suggestion is very different than "adding disk noises."   

The CRT monitor ratio in the DOS days was 4x3, whereas the de facto ratio today is 16x9.  This means that DOS games under DosBox don't fill the monitor.  Aesthetically it would be preferable to be able to fill that space with something other than black.

This wad mentioned in another thread as "a RetroArch thing" - it's actually a feature of almost all retro gaming emulators.  MAME, Daphne, RetroArch and many other emulators allow an arbitrary PNG (with transparency for allowing the game screen to shoe through) to be placed over the emulator's generated video.   This was suggested as a feature enhancement for DosBox only because Dosbox is currently in the minority.  

There is actually a fork of Dosbox (Dosbox-CRT) that does overlay a simulated monitor bezel.  Unfortunately the image is hard-coded and can't be changed.

**Comment by johnnovak** [comment]
> Aesthetically it would be preferable to be able to fill that space with something other than black.

We disagree on that; I prefer black.

Nothing has changed on my stance on this -- might come one day, but the priority is low, and I'm personally not interested in such features. Could be an optional feature one day though if it doesn't cause major conflicts with other more important features.

**Comment by aaciii** [comment]
Your emulator, your prerogative.  You might research why all other emulators have pursued this though - it's a feature that is used extensively, and there is a whole cottage industry around creating and packaging collections of screen bezels for various game content.

**Comment by aaciii** [comment]
> > Aesthetically it would be preferable to be able to fill that space with something other than black.
> 
> 
> 
> We disagree on that; I prefer black. I don't see why "every blank space" would need to be filled with "something".
> 
> 
> 
> Nothing has changed on my stance on this -- might come one day, but the priority is low, and I'm personally not interested in such features. Could be an optional feature one day though if it doesn't cause major conflicts with other more important features.
> 
> 
> 
> But a fixed image would already conflict with the auto-CRT shaders with integer scaling enabled. I guess you could use very specific bezel images with very specific scaling settings then (probably that's what the RetroArch folks do). Like I said, maybe one day.

It's more complex as the logic would be to apply the bezel image "on top" of the shader's output - the shader should not process the bezel image (eg, no scanlines on the bezel).   Could be done a few ways I suppose; there are reference implementetions that could be drawn from in other OSS projects that implement the feature as well as shaders (DOSbox-CRT, MAME, RetroArch, etc)

I get that it's not very interesting to you :)

**Comment by johnnovak** [comment]
> It's more complex as the logic would be to apply the bezel image "on top" of the shader's output

Nah, that's not complex at all. More like how it would interact with other features like I mentioned. A simplistic implementation would just load a static image and that's it, then you'd need to ensurethe emulated image ends up where it should to relative to the "hole" in the bezel.

**Comment by aaciii** [comment]
Yes that's one implementation.  Note that in some other implementations, the transparent "hole" in the PNG isn't symmetrical / rectangular - the PNG may actually (on purpose) overlay on top of the rendered emulation output.  Example:  ![image](https://github.com/user-attachments/assets/da4f3a8b-5b1b-4d80-ab53-5f1828a13418)

**Comment by johnnovak** [comment]
> So the emulator output needs to be confined to a viewport inside the transparency - but also, the shader needs to be in the video pipeline AFTER the emulator output and shader, but BEFORE application of the bezel. (Although obviously that's a corner case and usually the opening is screen-shaped). Examples:

Yeah I get all that, but I think you're overcomplicating the description. The transparent PNG would be applied on top of everything else (on top of the CRT shaded output), and job done. When simplified like this, this is more likely to happen (e.g. `bezel_image = blah.png` in the config).

What won't happen if the stuff I saw in some RA MegaBezelWhatever pack when some shader also applies reflections on the inner rim of the bezel image. That's a hard no, we won't go there (neither will we adapt the whole RA shader pipeline).

**Comment by aaciii** [comment]
> Yeah I get all that, but I think you're overcomplicating the description. The transparent PNG would be applied on top of everything else (on top of the CRT shaded output), and job done. Scaled to the output viewport (window, fulscreen) while preserving the aspect ratio of the image. When simplified like this, this is more likely to happen (e.g. `bezel_image = blah.png` in the config).

100% agree.   A simple "place the game output in the following coordinates" (x,y,width,height) and "place the full-screen PNG on top" is what most emulators provide.

> What won't happen if the stuff I saw in some RA MegaBezelWhatever pack when some shader also applies reflections on the inner rim of the bezel image. That's a hard no, we won't go there (neither will we adapt the whole RA shader pipeline).

Yeah that project is a bit of a "special case" - very few emulators support that.   It's a work in progress by a small team.
 
> PS: Although not my cup of tea, these images are pretty cool 👍🏻 😎 A useful use case could be to display the keymappings the game uses in the bezel area, which would be even handier if the bezel could be toggled via a hotkey.

Totally agree.   In many games (especially arcade style, where you have lots of buttons), knowing what buttons do what is critical.

**Comment by exoscoriae** [comment]
> > Yeah I get all that, but I think you're overcomplicating the description. The transparent PNG would be applied on top of everything else (on top of the CRT shaded output), and job done. Scaled to the output viewport (window, fulscreen) while preserving the aspect ratio of the image. When simplified like this, this is more likely to happen (e.g. `bezel_image = blah.png` in the config).
> 
> 100% agree. A simple "place the game output in the following coordinates" (x,y,width,height) and "place the full-screen PNG on top" is what most emulators provide.
> 
> > What won't happen if the stuff I saw in some RA MegaBezelWhatever pack when some shader also applies reflections on the inner rim of the bezel image. That's a hard no, we won't go there (neither will we adapt the whole RA shader pipeline).
> 
> Yeah that project is a bit of a "special case" - very few emulators support that. It's a work in progress by a small team.
> 
> > PS: Although not my cup of tea, these images are pretty cool 👍🏻 😎 A useful use case could be to display the keymappings the game uses in the bezel area, which would be even handier if the bezel could be toggled via a hotkey.
> 
> Totally agree. In many games (especially arcade style, where you have lots of buttons), knowing what buttons do what is critical.

I'm sorry your request was denied. I feel like as an option, it hurts no one who doesn't wish to use it. My project, eXoDOS, constantly has folks who request that I find a way to add bezels. It's not something I would likely use, but I would sure love to be able to offer it to those who are interested. These projects should be for everyone. Not personal playgrounds.

aacii, if you find a fork that enables this, please reach out to me and let me know. I am at a cross roads right now in converting my projects off of a now defunct fork of DOSBox, and would be very interested in a fork that is open minded, Thanks!

**Comment by johnnovak** [comment]
> I'm sorry your request was denied.

It wasn't. The ticket is still open, isn't it?

We agreed on a workable solution with @aaciii that I will implement (a simple PNG overlay) and he was happy with it. Read it the [guy's reply](https://github.com/dosbox-staging/dosbox-staging/issues/3558#issuecomment-2566566255) again, carefully:

> 100% agree. A simple "place the game output in the following coordinates" (x,y,width,height) and "place the full-screen PNG on top" is what most emulators provide.

> I feel like as an option, it hurts no one who doesn't wish to use it. 

As a non-developer, you have no idea how massively difficult supporting the full RA pipeline would be. You also don't know what we're planning on the shader front and how the RA pipeline would be incompatible with the auto-CRT stuff (without overcomplicating things massively).

> These projects should be for everyone. Not personal playgrounds.

That's your opinion. There is no "should" and "shouldn't" in open source. You get exactly what you pay for.

Everybody does what's fun _for them_ in their projects. Unless you're paying them to do what you want, or if you're working with a group of masochists, perhaps.

**Comment by exoscoriae** [comment]
I saw his reply.

I also know that staging doesn't currently do the simple "place the game output in the following coordinates"

You also made the executive decision to deny this request.

Closed or not, that is a final decisioin.

There is a big difference between a project that says, "We will consider this" and a project that says "I perosnally don't like this, so I'm denying it"

eXoDOS contains nearly 8,000 games. Many of them are games I have zero personal interest in. 

Some projects are personal playgrounds and some projects are focused more on the greater community. The reality is, bezels are a requested feature fairly often in my community. 

On one hand you claim I am not a developer, so I can't possibly understand how difficult it is, on the other you wrote above that it is "not that complex at all" and that is would be "simple" to place the output at a specific coordinates.

You have a habit of obfuscating things behind complexity when you have no desire to work on them.

**Comment by Grounded0** [comment]
There's already forks of DOSBox that implement bezels:

4:3

https://mattiasgustavsson.itch.io/dosbox-crt

16:9

https://nukekangaroo.itch.io/dosbox-crt-widescreen

Both are based on DOSBox 0.74, so I view them as technical dead ends by now. It's probably smarter to wait for us to implement those PNG bezels than to try to revive a dead DOSBox fork.

This ticket has been assigned low priority because stuff that directly negatively impacts gameplay, like our craptastic mapper, hasn't been rewritten yet. See how many gamers are directly negatively impacted by it right now:

- #3075

The time for this will come, but we need to get our technical debt down, which was inherited with this codebase. The forks that didn't try to get it down and instead added features first died.

DOSBox-X discussion over this:

- https://github.com/joncampbell123/dosbox-x/issues/3394

**Comment by johnnovak** [comment]
@exoscoriae Yes, initially I denied the request, the reconsidered... That happens. You can see that in the history. I'll give you that it's not immediately obvious unless you read the whole thread, but the "won't do" label has been removed and the ticket has been re-opened.

Another clue for the visual types:

<img width="286" height="83" alt="Image" src="https://github.com/user-attachments/assets/8f2b62dc-29a6-4a4e-a678-6a97d42203f0" />

> On one hand you claim I am not a developer, so I can't possibly understand how difficult it is, on the other you wrote above that it is "not that complex at all" and that is would be "simple" to place the output at a specific coordinates.

You're taking things out of context. The MegaBezel style implementation would be complex... (e.g., with supporting the emulated image being reflected on the bezel, just like MegaBezel does). A simplified PNG overlay implementation with the added ability to position the emulated video output would be, well, simple 😄 

https://github.com/dosbox-staging/dosbox-staging/issues/3558#issuecomment-2566146382

I'm giving you the benefit of doubt as you might not be familiar with the MegaBezel RA shader that many people are raving about. It's a super complicated monster, but sure, it looks impressive with the reflections and all if you're into that sort of thing:

https://forums.libretro.com/t/mega-bezel-reflection-shader-feedback-and-updates/25512

From memory, the top description went through a few iterations (it was more complex when the person initially raised it), but we don't have a history of that.

> You have a habit of obfuscating things behind complexity when you have no desire to work on them.

Or maybe that's just your impression without having a good grasp of the complexities and technicalities involved. Let's just say people having 20-30+ years of professional programming experience (most guys in our core team) and 3-5+ of specific DOSBox hacking experience are better situated to gauge what's difficult and what isn't...

Also, keep in mind, we only offer these explanations as a courtesy (instead of just simply saying "no" to a request).

TL;DR

- MegaBezel support, full RA pipeline — not coming
- Simple PNG overlays — I'm planning to implement it at some point. That's still the plan.

PS:

 > I also know that staging doesn't currently do the simple "place the game output in the following coordinates"

Yeah, that was the biggest complication, and once I came up with a workable solution which doesn't conflict with future goals (which exists in my head only, currently), I reopened the ticket. That was the biggest conceptual roadblock to supporting PNG overlays as I saw it.

**Comment by dux119-cmd** [comment]
I use eXoDOS with Staging under Batocera, and am looking forward to bezels. Not sure why the eXo project says this is denied when I see PNG bezels are in the pipe. I look forward to it when it comes ☺️ 👍

**Comment by johnnovak** [comment]
> Not sure why the eXo project says this is denied when I see PNG bezels are in the pipe.

You are as confused as I am about this 😂 

> I look forward to it when it comes ☺️ 👍

And come it will 😄 It will be a good "preparatory exercise" for the upcoming OSD work (the OSD will be "just" a different form of overlay).

**Comment by Torinde** [comment]
> I am at a cross roads right now in converting my projects off of a now defunct fork of DOSBox, and would be very interested in a fork that is open minded

Do you refer to ECE?

I fear exoscoriae-johnnovak misunderstanding is related to a Discord discussion about .conf files, resulting in feeling such as:

> You have a habit of obfuscating things behind complexity when you have no desire to work on them.

Which may be the answer why:

> Not sure why the eXo project says this is denied when I see PNG bezels are in the pipe.

Both eXo and Staging projects are fantastic and I hope you guys get again on the [same page](https://github.com/dosbox-staging/dosbox-staging/issues/3646) soon.

**Comment by johnnovak** [comment]
Noting down the solution ideas I've been thinking about on and off over the past 6 months or so:

- Add a single`overlay_image` setting to allow specifying a transparent PNG as the bezel/overlay image.
- The image will be composited on top of the shaded output; this way, one can use a PNG that darkens the edges of the emulated screen, or even adds a subtle white "gloss/reflection effect" to the emulated "glass" (if that's the person's thing 😄). You can also add a smudge layer, too 😎 Or blood, or whatever you enjoy splashing on your monitor... 😂 
- The image will be alwats sized so that it vertically fills the viewport. The width will be scaled proportionally so the image is not distorted. I think that's the most useful option so a single 4K image can be used for all resolutions down to 720p or whatever and any window size.
- One could use a 16:9 image, then if the window is not 16:9 aspect ratio, the "overhang" on the sides will be cut off (but that's fine).
- I've already added the `relative H% V%` option to the `viewport` setting to implement "CRT-like" stretch controls. With that, it's possible to size the image to be smaller than the full viewport, all relative to the size of the viewport. This way the user can freely set the size of the window; the video output will always appear correctly in relation to the overlay image, which is also sized relatively to the window.
- So far, this would only work if the "hole" in the overlay image is exactly centred for the emulated image to show through. So we'll also need a new `viewport_position` setting to support offsetting the image from the centre by a percentage amount (similarly to the `relative H% V%` option).
- The shaders will need an extra "geometry" parameter to distort the output to a screen-like curved surface (many CRT shaders already have this option, so it's just a matter of copying the code from somewhere).

I think relative positioning is best; you set up things once, then it works with any viewport size, screen resolution, etc. But maybe we could also support absolute positioning for completeness, although I don't see that as very useful (but it's simple to add, sure).

I'll need to play around with different types of relative percentage sizings—the goal is that it should just work relatively intuitively for overlay authors. It needs some experimentation, but conceptually it will be like relative percentage based sizings in CSS (as in web design).

The solution is an extension of the existing `viewport` sizing enhancements and doesn't really conflict with anything planned–that's the nice thing about it. I like things to work together logically as a system, not just a random collection of incompatible settings/ideas/whatever...

Here we go. Now everybody knows what's being planned 😄

**Comment by aaciii** [comment]
I love everything that you said except for one. One of the primary reasons for having this overlay capability is because DOS games are designed for 4 x 3 monitors, whereas the standard today is 16 x 9 monitors. Therefore it is very desirable to be able to put a 16 x 9 image on top, and have the black sides filled with some content Other than black. So I would not cut the sides of the image off, but rather render them.  the images above that I show as examples are 16 x 9 images on top of a 4 x 3 game screen.

**Comment by johnnovak** [comment]
> I love everything that you said except for one. One of the primary reasons for having this overlay capability is because DOS games are designed for 4 x 3 monitors, whereas the standard today is 16 x 9 monitors. Therefore it is very desirable to be able to put a 16 x 9 image on top, and have the black sides filled with some content Other than black. So I would not cut the sides of the image off, but rather render them. the images above that I show as examples are 16 x 9 images on top of a 4 x 3 game screen.

That's exactly how it will work, you misunderstood me slighthly 😄 I was just describing what would happen if the user starts resizing a say 16:9 AR window closer to 4:3 AR -- then the sides of the 16:9 AR overlay image will be clipped (instead of say the image being fit horizontally or stretched to the available width).

It will work how you want it to work 😄

**Comment by johnnovak** [comment]
Curvature code by Hyllian (in the pass0 and pass1 files)

https://gitlab.incom.co/libretro/slang-shaders/-/tree/3796d4448201e34f6f240a001b9db1c4c34f606b/crt/shaders/hyllian/

We'll need to add this to the CRT shaders as well.

**Comment by johnnovak** [comment]
The Duimon Mega Bezel pack contains a bunch of interesting background and bezel images. Will come in handy during feature development.

IBM monitor bezels:

https://github.com/Duimon/Duimon-Mega-Bezel/tree/main/Graphics/IBM_5150

Background images:

https://github.com/Duimon/Duimon-Mega-Bezel/tree/main/Graphics/_Backgrounds
